lumeo_api_client/src/error.rs

```diff
@@ -2,31 +2,55 @@ use std::fmt;
 
 use reqwest::{Method, Response, StatusCode};
 use serde::{Deserialize, Serialize};
-use strum::EnumString;
+use strum::{EnumString, EnumVariantNames, VariantNames};
 use thiserror::Error;
 
 use crate::Result;
 
-#[derive(EnumString, Error, Debug)]
+#[derive(EnumVariantNames, Error, Debug)]
 #[strum(serialize_all = "kebab-case")]
 pub enum ApiError {
     #[error("The gateway this access token belonged to has been deleted (`gateway-deleted`)")]
     GatewayDeleted,
     #[error("User credentials are invalid (`invalid-credentials`)")]
     InvalidCredentials,
-    #[error("Resource not found (`resource-not-found`)")]
-    ResourceNotFound,
+    #[error("Resource not found (`resource-not-found`), resource: {0}")]
+    ResourceNotFound(#[source] ResourceNotFound),
     #[doc(hidden)]
     #[strum(disabled)]
     #[error("{message} (`{code}`)")]
     Other { code: String, message: String },
 }
 
+#[derive(Debug, Error, EnumString)]
+pub enum ResourceNotFound {
+    #[error("Deployment")]
+    #[strum(serialize = "deployment")]
+    DeploymentNotFound,
+    #[doc(hidden)]
+    #[strum(disabled)]
+    #[error("Other({0})")]
+    Other(String),
+}
+
+fn try_resource(value: serde_json::Value) -> Option<ResourceNotFound> {
+    value
+        .as_object()
+        .and_then(|obj| obj.values().next())
+        .and_then(|resource| resource.as_str())
+        .and_then(|it| {
+            it.parse::<ResourceNotFound>()
+                .unwrap_or_else(|_| ResourceNotFound::Other(it.to_owned()))
+                .into()
+        })
+}
```

Comment:
```suggestion
impl ResourceNotFound {
    fn from_context(mut context: serde_json::Value) -> Option<Self> {
        let resource = match context.as_object_mut()?
            .remove("resource")? {
            serde_json::Value::String(resource) => resource,
            _ => return None,
        };

        let resource_not_found = resource.parse::<ResourceNotFound>()
            .unwrap_or_else(|_| ResourceNotFound::Other(resource));
        Some(resource_not_found)
    }
}
```

You could also use the following if you don't mind the string copy of the `to_owned`:

```rust
impl ResourceNotFound {
    fn from_context(context: &serde_json::Value) -> Option<Self> {
        let resource = context.as_object()?.get("resource")?.as_str()?;

        let resource_not_found = resource.parse::<ResourceNotFound>()
            .unwrap_or_else(|_| ResourceNotFound::Other(resource.to_owned()));
        Some(resource_not_found)
    }
}

```